S3Source.cc

```diff
@@ -0,0 +1,248 @@
+    if ( cls == nullptr ) {
+      throw std::runtime_error("No TClass reflection available for " + pi.productname());
+    }
+    dataBuffers_[i] = cls->New();
+    dataProducts_.emplace_back(i, &dataBuffers_[i], pi.productname(), cls, &delayedRetriever_);
+    deserializers_.emplace_back(cls);
+    ++i;
+  }
+}
+
+S3Source::LaneInfo::~LaneInfo() {
+  auto it = dataProducts_.begin();
+  for( void * b: dataBuffers_) {
+    it->classType()->Destructor(b);
+    ++it;
+  }
+}
+
+std::pair<const char*, size_t> DelayedProductStripeRetriever::bufferAt(size_t globalEventIndex) const {
+  std::call_once(flag_, [this](){
+    conn_->get(name_, [this](S3Request* req) {
+        if ( req->status == S3Request::Status::ok ) {
+          if ( not data_.ParseFromString(req->buffer) ) {
+            throw std::runtime_error("Could not deserialize ProductStripe for key " + name_);
+          }
+        }
+        else { throw std::runtime_error("Could not retrieve ProductStripe for key " + name_); }
+      });
+    });
+  assert(globalOffset_ <= globalEventIndex);
+  size_t offset = globalEventIndex - globalOffset_;
+  assert(offset < data_.offsets_size());
+  size_t bstart = (offset == 0) ? 0 : data_.offsets(offset-1);
+  size_t bstop = data_.offsets(offset);
+  return {&data_.content()[bstart], bstop - bstart};
+}
+
+size_t S3Source::numberOfDataProducts() const {
+  return laneInfos_[0].dataProducts_.size();
+}
+
+std::vector<DataProductRetriever>& S3Source::dataProducts(unsigned int iLane, long iEventIndex) {
+  return laneInfos_[iLane].dataProducts_;
+}
+
+EventIdentifier S3Source::eventIdentifier(unsigned int iLane, long iEventIndex) {
+  return laneInfos_[iLane].eventID_;
+}
+
+void S3Source::readEventAsync(unsigned int iLane, long iEventIndex, OptionalTaskHolder iTask) {
+  queue_.push(*iTask.group(), [iLane, optTask = std::move(iTask), this]() mutable {
+      auto start = std::chrono::high_resolution_clock::now();
+      if(
+          (nextEventStripe_ < index_.packedeventstripes_size())
+          or (nextEventInStripe_ < currentEventStripe_.events_size())
+        )
+      {
+        // default-constructed currentEventStripe_ will have size zero, so 0, 0 will load first stripe
+        if(nextEventInStripe_ == currentEventStripe_.events_size()) {
+          // need to read ahead
+          // TODO: compression
+          currentEventStripe_.ParseFromString(index_.packedeventstripes(nextEventStripe_++));
+          nextEventInStripe_ = 0;
+        }
+        const auto event = currentEventStripe_.events(nextEventInStripe_);
+        if ( verbose_ >= 2 ) std::cout << event.DebugString() << "\n";
+        size_t globalEventIndex = event.offset();
+        laneInfos_[iLane].eventID_.run = event.run();
+        laneInfos_[iLane].eventID_.lumi = event.lumi();
+        laneInfos_[iLane].eventID_.event = event.event();
+        auto stripes = std::vector<std::shared_ptr<const DelayedProductStripeRetriever>>();
```

Comment:
The idea here was to detach the lifetime of the input buffer, but it does mean making lots of copies of shared pointers. Not sure how bad that is